qi-lib/flow/compiler.rkt

```diff
@@ -424,9 +424,8 @@ the DSL.
        #'(qi0->racket ground)]
       [(_ [cond:clause body:clause])
        #'(qi0->racket (~> (pass cond) body))]
-      [(_ [cond:clause body:clause]  ...+)
-       #:with c+bs #'(list (cons (qi0->racket cond) (qi0->racket body)) ...)
-       #'(qi0->racket (~>> (partition-values c+bs)))]))
+      [(_ [cond:clause body:clause] [conds:clause bodies:clause] ...+)
+       #'(qi0->racket (sieve cond body (partition [conds bodies] ...)))]))
```

Comment:
I'll be curious to see how expressively the optimizing compiler can detect the expansion to rewrite back to `partition-values` because of the nesting. For example:

```racket
(sieve a b
  (sieve c d
    (sieve e f
      (~> (pass g) h))))
```
becomes (roughly)
```
(partition-values
  (list
    (cons a b) (cons b c) (cons e f) (cons g h)))
```

Reply:
Is that `(cons b c)` supposed to be `(cons c d)`? But yeah, I don't really know how the optimizations will be implemented yet. Should be interesting 😝

Reply:
Yeah, fixed.

I have no suggestions—on the one hand, it's probably a "classic" recursive algorithm. OTOH, the pattern-matching transformation was concise :)